src/rulesets/Base/indexing.jl

```diff
@@ -131,6 +131,7 @@ allow `eltype(dy)`, nor does it work for many structured matrices.
 """
 _setindex_zero(x::AbstractArray{<:Number}, dy, inds::Integer...) = fill!(similar(x, typeof(dy), axes(x)), false)
 _setindex_zero(x::AbstractArray{<:Number}, dy, inds...) = fill!(similar(x, eltype(dy), axes(x)), false)
```

Comment:
Is there a reason for why these are not just
```suggestion
_setindex_zero(x::AbstractArray{<:Number}, dy, inds::Integer...) = fill!(similar(x, typeof(dy)), false)
_setindex_zero(x::AbstractArray{<:Number}, dy, inds...) = fill!(similar(x, eltype(dy)), false)
```
AFAICT this would also fix the AxisArrays problem.

Reply:
This breaks existing tests. The problem is if you don't pass the axes, then you don't get a dense array.

Reply:
Which tests are broken? The two-arg method is even advised in the Julia docs: https://docs.julialang.org/en/v1/manual/methods/#Building-a-similar-type-with-a-different-type-parameter

Reply:
The 3-arg one removes structured matrices like Symmetric, iirc

Reply:
Maybe we should just add special cases for these? At first glance, it doesn't seem very desirable to remove structure (as the AxisArrays case shows).

Reply:
IMO the ideal situation is for `axes` to return more information in AxisArrays, alla https://github.com/mcabbott/AxisKeys.jl/pull/6 , since the relevant properties do belong to individual axes, not to the whole (like Symmetric). But we ran out of energy to fix things.